tests/test_site_apportionment/test_site_apportionment.py

```diff
@@ -58,3 +67,325 @@ def test_create_notnull_mask(self):
+
+def create_sample_nan_postcodes_df():
+
+    sample_cols = ["reference", "period", "601", "headcount_dummy", "211_dummy"]
+
+    nan_as_str = str(np.nan)
+
+    sample_data =    [
+        [4990000126, 202212, "IJ5 6KL", 3877, 5808144],
+        [4990000126, 202212, np.nan,        3877, 5808144], # Numpy nan as postcode
+        [4990000126, 202212, nan_as_str,     3877, 5808144], # string representation of numpy nan as postcode
+        [4990000126, 202212, None,           3877, 5808144], # None as postcode
+        ]
+
+    # Create the DataFrame
+    return pd.DataFrame(data=sample_data, columns=sample_cols)
+
+def test_count_unique_postcodes_in_col_nan_postcodes():
+    """This test checks that the function count_unique_postcodes_in_col can handle NaN postcodes.
+
+    The NaNs are represented by np.nan and the string representation of np.nan, 'nan'.
+
+    As the data is read from a file, received from SPP, missing values may be
+     represented as the string 'nan' so this might be worth testing for.
+    """
+
+    sample_df = create_sample_nan_postcodes_df()
+
+    # Apply the function
+    result_df = count_unique_postcodes_in_col(sample_df)
+
+    # Expected output
+    expected_cols = ['reference', 'period', '601', "headcount_dummy", "211_dummy", '601_count']
+
+    nan_as_str = str(np.nan)
+
+    expected_data = [
+        [4990000126, 202212, "IJ5 6KL", 3877, 5808144, 1],
+        [4990000126, 202212, np.nan,        3877, 5808144, 1],
+        [4990000126, 202212, nan_as_str,     3877, 5808144, 1],
+    ]
+
+    # Make sure the expected data is in a DataFrame
+    expected_df = pd.DataFrame(data=expected_data, columns=expected_cols)
+
+    # Check if the output is as expected
+    pd.testing.assert_frame_equal(result_df, expected_df, check_dtype=False)
+
+
+def create_sample_missing_columns_df():
+    """
+    Creates a sample DataFrame using predefined data and columns.
+
+    Returns:
+        pd.DataFrame: The created DataFrame.
+    """
+    # Create the columns missing the postcode column "601"
+    sample_cols = ["reference", "period", "headcount_dummy", "211_dummy"]
+
+    # This data is missing the postcode column
+    sample_data = [
+    [4990000000, 202212, 5663, 8855030],
+    [4990000000, 202212, 242, 5949501],
+    [4990000084, 202212, 8020, 5085659],
+    [4990000126, 202212, 3877, 5808144],
+    [4990000126, 202212, 2756, 8889091],
+    [4990000126, 202212, 8832, 4722201],
+    [4990000252, 202212, 7084, 7058606],
+    [4990000294, 202212, 3188, 1828472],
+    [4990000336, 202212, 62, 3262548],
+    [4990000378, 202212, 1180, 9348647]
+    ]
+
+    return pd.DataFrame(data=sample_data, columns=sample_cols)
+
+
+def test_count_unique_postcodes_in_col_missigng_col():
+    """Checks that an error is raised when the postcode column is missing."""
+
+    df = create_sample_missing_columns_df()
+
+    # Apply the function
+    with pytest.raises(KeyError):
```

Comment:
This is throwing up a similar issue, where I've changed the name of the col in the test to 601 and it is still passing